cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -269,6 +278,60 @@ class device_row_comparator {
   std::optional<device_span<null_order const>> const _null_precedence;
 };  // class device_row_comparator
 
+/**
+ * @brief Wraps and interprets the result of device_row_comparator, true if the result is
+ * weak_ordering::LESS meaning one row is lexicographically *less* than another row.
+ *
+ * @tparam Nullate A cudf::nullate type describing whether to check for nulls.
+ */
+template <typename Nullate>
+class device_less_comparator {
+  friend class self_comparator;
+  /**
+   * @brief Construct a function object for performing a lexicographic
+   * comparison between the rows of two tables.
+   *
+   * @param has_nulls Indicates if either input table contains columns with nulls.
+   * @param lhs The first table
+   * @param rhs The second table (may be the same table as `lhs`)
+   * @param depth Optional, device array the same length as a row that contains starting depths of
+   * columns if they're nested, and 0 otherwise.
+   * @param column_order Optional, device array the same length as a row that indicates the desired
+   * ascending/descending order of each column in a row. If `nullopt`, it is assumed all columns are
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If `nullopt`, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   */
+  device_less_comparator(
+    Nullate check_nulls,
+    table_device_view lhs,
+    table_device_view rhs,
+    std::optional<device_span<int const>> depth                  = std::nullopt,
+    std::optional<device_span<order const>> column_order         = std::nullopt,
+    std::optional<device_span<null_order const>> null_precedence = std::nullopt) noexcept
+    : comparator{check_nulls, lhs, rhs, depth, column_order, null_precedence}
+  {
+  }
+
+ public:
+  /**
+   * @brief Checks whether the row at `lhs_index` in the `lhs` table compares
+   * lexicographically less than the row at `rhs_index` in the `rhs` table.
+   *
+   * @param lhs_index The index of row in the `lhs` table to examine
+   * @param rhs_index The index of the row in the `rhs` table to examine
+   * @return `true` if row from the `lhs` table compares less than row in the `rhs` table
+   */
+  __device__ bool operator()(size_type const lhs_index, size_type const rhs_index) const noexcept
+  {
+    return comparator(lhs_index, rhs_index) == weak_ordering::LESS;
```

Comment:
Knowing that we'll want other things like `device_less_equivalent_comparator`, I'd make some kind of base `_impl` for checking the result of `device_row_comparator` against some specified value(s).

```
template <typename Nullate, weak_ordering... values>
struct weak_ordering_comparator_impl{
   bool operator()(...){
      auto const result = comparator(lhs, rhs);
      return (result == values) || ...
   }
};

template <typename Nullate>
using less_comparator = weak_ordering_comparator_impl<Nullate, weak_ordering::LESS>;

template <typename Nullate>
using less_equivalent_comparator = weak_ordering_comparator_impl<Nullate, weak_ordering::LESS, weak_ordering::EQUIVALENT>;
```

Reply:
Yes! I was trying to decide how that should look, and I really like this proposal.

Reply:
Godbolt example that fleshes out the above concept: https://godbolt.org/z/PGenKvKG8

@jrhemstad Happy to hear critique of this. I'm discussing with @rwlee now. Is there a way to avoid recursion when handling `values...`? Is `(result == values) || ...` supposed to be valid syntax?

Reply:
> Is (result == values) || ... supposed to be valid syntax?

Kind of, though I didn't explicitly test it. The idea is to use a [fold expression](https://en.cppreference.com/w/cpp/language/fold)

Reply:
Updated your example to use a fold expression. My example above was missing an extra set of parentheses

https://godbolt.org/z/eTqMvz9T6

Reply:
That's so beautiful! I had no idea that you could use folds with multiple operators based on precedence like that. I've only used trivial folds. Thank you!

Reply:
Oh that's awesome, I assumed the `...` were places where you wanted me to fill in holes. I've adjusted the code and am rebuilding.

Reply:
> I assumed the ... were places where you wanted me to fill in holes.

Kind of! They are the places I want the _compiler_ to fill in the holes :)